kong/mlcache.lua

```diff
@@ -0,0 +1,466 @@
+
+    local ok, err = self.dict:set(key, shm_marshalled, ttl)
+    if not ok then
+        return nil, "could not write to lua_shared_dict: " .. err
+    end
+
+    -- set our own worker's LRU cache
+
+    return set_lru(self, key, value, ttl)
+end
+
+
+local function unlock_and_ret(lock, res, err)
+    local ok, lerr = lock:unlock()
+    if not ok then
+        return nil, "could not unlock callback: " .. lerr
+    end
+
+    return res, err
+end
+
+
+function _M:get(key, opts, cb, ...)
+    if type(key) ~= "string" then
+        return error("key must be a string")
+    end
+
+    if type(cb) ~= "function" then
+        return error("callback must be a function")
+    end
+
+    -- opts validation
+
+    local ttl
+    local neg_ttl
+
+    if opts then
+        if type(opts) ~= "table" then
+            return error("opts must be a table")
+        end
+
+        if opts.ttl then
+            if type(opts.ttl) ~= "number" then
+                return error("opts.ttl must be a number")
+            end
+
+            if opts.ttl < 0 then
+                return error("opts.ttl must be >= 0")
+            end
+        end
+
+        if opts.neg_ttl then
+            if type(opts.neg_ttl) ~= "number" then
+                return error("opts.neg_ttl must be a number")
+            end
+
+            if opts.neg_ttl < 0 then
+                return error("opts.neg_ttl must be >= 0")
+            end
+        end
+
+        ttl     = opts.ttl
+        neg_ttl = opts.neg_ttl
+    end
+
+    if not ttl then
+        ttl = self.ttl
+    end
+
+    if not neg_ttl then
+        neg_ttl = self.neg_ttl
+    end
+
+    -- worker LRU cache retrieval
+
+    local data = self.lru:get(key)
+    if data == CACHE_MISS_SENTINEL_LRU then
+        return nil
+    end
+
+    if data ~= nil then
+        return data
+    end
+
+    -- not in worker's LRU cache, need shm lookup
+
+    local err
+    data, err = shmlru_get(self, key)
+    if err then
+        return nil, err
+    end
+
+    if data == CACHE_MISS_SENTINEL_LRU then
+        return nil
+    end
+
+    if data ~= nil then
+        return data
+    end
+
+    -- not in shm either
+    -- single worker must execute the callback
+
+    local lock, err = resty_lock:new(self.shm)
+    if not lock then
+        return nil, "could not create lock: " .. err
+    end
+
+    local elapsed, err = lock:lock(LOCK_KEY_PREFIX .. key)
+    if not elapsed then
+        return nil, "could not aquire callback lock: " .. err
+    end
+
+    -- check for another worker's success at running the callback
+
+    data, err = shmlru_get(self, key)
+    if err then
+        return unlock_and_ret(lock, nil, err)
+    end
+
+    if data then
+        if data == CACHE_MISS_SENTINEL_LRU then
+            return unlock_and_ret(lock, nil)
+        end
+
+        return unlock_and_ret(lock, data)
+    end
+
+    -- still not in shm, we are responsible for running the callback
+
+    local ok, err = pcall(cb, ...)
```

Comment:
This is why there was a timeout value in the original implementation. In the current one with a double lock mechanism. there is a timeout for both locks as well, in case such a scenario happens.